temporarily prevent Copyable types from using `consuming` and `borrowing`

kavon opened the pull request:
There is a strong desire for "no-implicit-copy" semantics to be applied to Copyable values marked with the newer `borrowing` and `consuming` ownership specifiers. If we lock ourselves into what SE-390 specifies now for Copyable types in the implementation, then we'll have to introduce a source break when we want such values to have the desired semantics.

Before there's wider adoption, I'm making it an error to use those newer names. If for some reason this is enabling a critical optimization for your use case, please use the old names instead:

 - `__owned` for `consuming` parameters
 - `__consuming` for `consuming` methods
 - `__shared` for `borrowing` parameters

 NOTE: the older names have their ownership attribute mangled into
 the name of the function itself, so there's a possibility of ABI
 breaks in the future if you move from the old name to the new one.
 There is some consideration being made to allow for a migration to
 the new names in the future without breaking ABI, so do reach out if
 this is an issue for you.

 resolves rdar://108538971

kavon commented:
@swift-ci please test

gottesmm commented:
@kavon this looks good to me. Except, can you add a flag that disables this for testing cases. Or change the tests that use this to use __shared/etc. Personally, I would rather the tests just have a special flag that way we future proof and don't have to change them in the future.

kavon committed:
add and update tests for Copyable types not having ownership modifiers

part of resolving rdar://108538971

kavon committed:
temporarily prevent Copyable types from using `consuming` and `borrowing`

There is a strong desire for "no-implicit-copy" semantics to be
applied to Copyable values marked with the newer `borrowing` and
`consuming` ownership specifiers. If we lock ourselves into what
SE-390 specifies now for Copyable types in the implementation, then
we'll have to introduce a source break when we want such values to
have the desired semantics.

Before there's wider adoption, I'm making it an error to use those
newer names. If for some reason this is enabling a critical
optimization for your use case, please use the old names instead:

 - `__owned` for `consuming` parameters
 - `__consuming` for `consuming` methods
 - `__shared` for `borrowing` parameters

 NOTE: the older names have their ownership attribute mangled into
 the name of the function itself, so there's a possibility of ABI
 breaks in the future if you move from the old name to the new one.
 There is some consideration being made to allow for a migration to
 the new names in the future without breaking ABI, so do reach out if
 this is an issue for you.

 resolves rdar://108538971

kavon commented on test/attr/lexical.swift:
```diff
@@ -44,13 +44,13 @@ func foo() {
 func zoo(@_eagerMove _ : consuming MoveOnly) {} // expected-error {{@_eagerMove cannot be applied to NonCopyable types}}
 
-func zooo(@_noEagerMove  _ : consuming C) {} // ok, only way to spell this behavior
+func zooo(@_noEagerMove  _ : C) {} // ok, only way to spell this behavior
```
@nate-chandler does this still test what it's meant to?

nate-chandler replied:
No, it's supposed to verify that it's legal to decorate a Copyable argument with both `@_noEagerMove` and `consuming`.  Since this PR bans-for-now consuming Copyable arguments, that case doesn't exist anymore.  Ideally this test would be commented out with a TODO, but this is fine.

kavon replied:
I went with the TODO!

kavon commented:
@gottesmm I went with guarding the ability to write it behind `NoImplicitCopy`.

kavon commented:
@swift-ci please test

kavon commented:
@swift-ci please smoke test

kavon commented:
@swift-ci please smoke test
